bsip-0061.md

```diff
@@ -0,0 +1,182 @@
+
+    - to allow users to unlock their accounts;
+    - to prevent circular dependencies.
+
+This solution allows:
+
+    - to resolve user complaints;
+    - to increase confidence in BitShares.
+
+# Rationale
+## Prevent circular dependencies
+To prevent circular dependencies account evaluators should contain verificaton code that detects cycled authorities. The new function `verify_cycled_authority()` traverses down the hierarchy of accounts and checks if the transaction can be potentially signed with this account. To avoid any misunderstanding of existing authorization mechanism the `sign_state` class will be reused. Parametrizing this class with `verify_only`-flag helps to do code of `verify_cycled_authority()` simple, readable and reliable.
+
+## "Lock"-operation
+"Lock"-operation is the last *account_update_operation* that led the account to a locked state. No more operations possible on behalf of this account.
+
+## Unlocking account
+To unlock account we have to find previous *account_create_operation* (or *account_update_operation*) that changed owner or active authority. This operation will be reapplied to undo "lock"-operation if possible. See Discussion.
+
+## Fix locked accounts
+At the beginning find all locked accounts in first maintenance after hard-fork. Maybe locked accounts should be marked for unlock. Then fix them. There are two approaches here.
+
+First - automatic (unconditional)
+
+    - Find latest operation ("lock"-operation) among operations of cycled accounts. Do it for each cycle.
+    - Undo all "lock"-operations. Find previous account update/create operation that changes authorities and redo it.
+
+Second - manual (on-demand). Unlocking can be performed only by the account that performed the operation that led to the blocking, using the credentials immediately prior to blocking. That is, not all the keys from the history of this account fit. Thus, you can unlock an account that has been blocked only by mistake and only if the necessary credentials are available. Here we will avoid illegal behavior.
+
+    - New type of operation defined: unlock_account_operation { account_to_unlock : account_id_type }.
+    - Initiator (potentially *account_to_unlock*) signs the transaction using its previous authority preceding the locked state.
+    - Initiator (potentially *account_to_unlock*) pays fee for this operation.
+    - *account_to_unlock* should have sufficient balances to perform this operation.
+    - Check if *account_to_unlock* is locked in unlock_account_evaluator.
+    - Check if *account_to_unlock* is authorized to perform the operation (transaction signed with previous authority).
+    - Find previous account update/create operation that changes authorities - restore point.
+    - Undo "lock"-operation: extract owner and active authorities from restore point, update account with extracted data.
+    - Push corresponding virtual account_update_operation.
+
+# Specifications
+## Prevent circular dependencies
```

Comment:
The specification is incomplete. It does not specify the unlocking process.

Also, please keep in mind that the target audience of a BSIP consists largely of non-programmers. Code as specification is insufficient. Pseudocode is welcome for data structures etc., but logic should be describe in a human-readable way. (Some of the "Rationale" can be included in the specification.)